spec.emu

```diff
@@ -187,33 +1022,48 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
         </emu-alg>
         <emu-grammar>ModuleItemList : ModuleItemList ModuleItem</emu-grammar>
         <emu-alg>
-          1. Let _moduleNames_ be ModuleRequests of |ModuleItemList|.
-          1. Let _additionalNames_ be ModuleRequests of |ModuleItem|.
-          1. Append to _moduleNames_ each element of _additionalNames_ that is not already an element of _moduleNames_.
-          1. Return _moduleNames_.
+          1. Let <del>_moduleNames_</del><ins>_requests_</ins> be ModuleRequests of |ModuleItemList|.
+          1. Let <del>_additionalNames_</del><ins>_additionalRequests_</ins> be ModuleRequests of |ModuleItem|.
+          1. <del>For each String _name_ of _additionalNames_, do</del>
+          1. <ins>For each ModuleRequest Record _mr_ of _additionalRequests_, do</ins>
+            1. <ins>Let _found_ be *false*.</ins>
+            1. <ins>For each ModuleRequest Record _mr2_ of _requests_, do</ins>
+              1. <ins>If _mr_.[[Specifer]] is _mr2_.[[Specifer]] and _mr_.[[Phase]] is _mr2_.[[Phase]], then</ins>
+                1. <ins>Assert: _found_ is *false*.</ins>
+                1. <ins>Set _found_ to *true*.</ins>
+            1. <del>If _moduleNames_ does not contain _name_, then</del>
+            1. <ins>If _found_ is *false*, then</ins>
+              1. Append <del>_name_</del><ins>_mr_</ins> to <del>_moduleNames_</del><ins>_requests_</ins>.
```

Comment:
Since `ModuleRequests` is used to populate the `[[RequestedModules]]` field of module records, and since we want to load each module once even if it has different phases, we should deduplicate this list. If two modules have the same specifier but different phases, we should still set _found_ to true (so that don't we append it again) and just upgrade the phase in the existing ModuleRequest Record to max(phase1, phase2).

I believe that the reason you removed the de-duplication logic is because we still want the duplicates in ImportEntries, because if we have `import source s from "./a.js"; import d from "./a.js"` we want to bind both `s` and `d`. It's true that ImportEntries internally calls ModuleRequests, however:
- it only calls ModuleRequests on pieces of individual import statements, and thus it wouldn't go through this deduplication logic. It's similar to how in
  ```js
  import a from "./a.js";
  import b from "./a.js";
  ```
  we have two different ImportEntries but only one ModuleRequest.
- if necessary to make it more clear, you could make it not use ModuleRequests anymore and instead do something like https://nicolo-ribaudo.github.io/proposal-defer-import-eval/#sec-static-semantics-importentries